pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+func (sasr *SummaryAllocationSetRange) AggregateBy(aggregateBy []string, options *AllocationAggregationOptions) error {
+	sasr.Lock()
+	defer sasr.Unlock()
+
+	for _, sas := range sasr.SummaryAllocationSets {
+		err := sas.AggregateBy(aggregateBy, options)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Append appends the given AllocationSet to the end of the range. It does not
+// validate whether or not that violates window continuity.
+func (sasr *SummaryAllocationSetRange) Append(sas *SummaryAllocationSet) error {
+	if sasr.Step != 0 && sas.Window.Duration() != sasr.Step {
+		return fmt.Errorf("cannot append set with duration %s to range of step %s", sas.Window.Duration(), sasr.Step)
+	}
+
+	sasr.Lock()
+	defer sasr.Unlock()
+
+	// Append to list of sets
+	sasr.SummaryAllocationSets = append(sasr.SummaryAllocationSets, sas)
+
+	// Set step, if not set
+	if sasr.Step == 0 {
+		sasr.Step = sas.Window.Duration()
+	}
+
+	// Adjust window
+	if sasr.Window.Start() == nil || (sas.Window.Start() != nil && sas.Window.Start().Before(*sasr.Window.Start())) {
+		sasr.Window.start = sas.Window.Start()
+	}
+	if sasr.Window.End() == nil || (sas.Window.End() != nil && sas.Window.End().After(*sasr.Window.End())) {
+		sasr.Window.end = sas.Window.End()
+	}
+
+	return nil
+}
+
+// Each invokes the given function for each AllocationSet in the range
+func (sasr *SummaryAllocationSetRange) Each(f func(int, *SummaryAllocationSet)) {
+	if sasr == nil {
+		return
+	}
+
+	for i, as := range sasr.SummaryAllocationSets {
+		f(i, as)
+	}
+}
+
+// TODO this stinks. Can we do better with external cost so that we can remove?
+func (sasr *SummaryAllocationSetRange) InsertExternalAllocations(that *AllocationSetRange) error {
+	if sasr == nil {
+		return fmt.Errorf("cannot insert range into nil AllocationSetRange")
+	}
+
+	// keys maps window to index in asr
+	keys := map[string]int{}
+	for i, as := range sasr.SummaryAllocationSets {
+		if as == nil {
+			continue
+		}
+		keys[as.Window.String()] = i
+	}
+
+	// Nothing to merge, so simply return
+	if len(keys) == 0 {
+		return nil
+	}
+
+	var err error
+	that.Each(func(j int, thatAS *AllocationSet) {
+		if thatAS == nil || err != nil {
+			return
+		}
+
+		// Find matching AllocationSet in asr
+		i, ok := keys[thatAS.Window.String()]
+		if !ok {
+			err = fmt.Errorf("cannot merge AllocationSet into window that does not exist: %s", thatAS.Window.String())
+			return
+		}
+		sas := sasr.SummaryAllocationSets[i]
+
+		// Insert each Allocation from the given set
+		thatAS.Each(func(k string, alloc *Allocation) {
+			externalSA := NewSummaryAllocation(alloc, true, true)
+			err = sas.Insert(externalSA)
+			if err != nil {
+				// TODO ?
+				log.Errorf("SummaryAllocation: error inserting %s: %s", k, err)
+			}
+		})
+	})
+
+	// err might be nil
+	return err
+}
+
+// TODO Custom MarshalJSON and UnmarshalJSON for these?
+// - Step is uintelligible (microseconds??)
+// - TotalCost would be nice
```

Comment:
Not sure about the costs of doing so, but we could separate the types between `SummaryAllocation` (used internally to actually compute the summarized allocation) and `SummaryAllocationResponse` which is easy to serialize/deserialize based on struct tags. Then, instead of writing a custom `MarshalJSON`, you just have to write a method like 
```go
func (sa *SummaryAllocation) ToResponse() SummaryAllocationResponse
```

Reply:
True. For now, I think we don't have an issue, but I'm curious to hear your vision for a `...Response` type. Would we make one for each thing we JSON-serialize, ideally? What are the pros/cons?

Reply:
My ideal world does include fully separating API response types (serialized) and internal data types. This relates to my earlier comment about (de)normalization.

The main con I see is overhead in converting between types. The main pro I see is greater flexibility in internal data representation while maintaining API contracts by explicitly fencing the two areas off in the type system.